core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -380,6 +381,7 @@ public CompletableFuture<Void> maybeInitialize() {
         // All the pending requests should wait to get completed before the share partition is initialized.
         // Attain lock to avoid any concurrent requests to be processed.
         lock.writeLock().lock();
+        AtomicReference<Optional<Throwable>> futureException = new AtomicReference<>(Optional.empty());
         try {
             // Re-check the state to verify if previous requests has already initialized the share partition.
             maybeCompleteInitialization(future);
```

Comment:
It seems to me that `maybeCompleteInitialization` should not complete the future under the write lock either. It could return an exception which is then places into the `futureException` and then the future is completed in the finally block.

Reply:
agreed